### PR TITLE
MW-100: Allow Dynamic Images in Jasper Reports

### PR DESCRIPTION
**Pull request by jkondrat**

### CODE REVIEW

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java [hunk 1]
```diff
@@ -82,6 +87,14 @@
   @Setter
   private List<JasperTemplateParameter> templateParameters;
 
+  @ManyToMany(fetch = FetchType.EAGER)
+  @JoinTable(name = "jasper_templates_report_images",
+      joinColumns = @JoinColumn(name = "jaspertemplatesid", nullable = false),
```
Comment (palbecki) [inline src/main/java/org/openlmis/report/domain/JasperTemplate.java]: shouldn't this be named jaspertemplateid?

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java [hunk 2]
```diff
@@ -82,6 +87,14 @@
...
+      inverseJoinColumns = @JoinColumn(name = "reportimagesid", nullable = false))
```
Comment (palbecki) [inline src/main/java/org/openlmis/report/domain/JasperTemplate.java]: shouldn't this be named reportimageid?

File: src/main/java/org/openlmis/report/domain/ReportImage.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    return jasperTemplate;
+  }
+
+  /**
+   * Copy values of attributes into new or updated ReportImage.
+   *
+   * @param jasperTemplate Template with new values.
+   */
+  public void updateFrom(ReportImage jasperTemplate) {
```
Comment (palbecki) [inline src/main/java/org/openlmis/report/domain/ReportImage.java]: wrong param name

File: src/main/java/org/openlmis/report/service/ReportImageService.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+@Service
+@SuppressWarnings("PMD.TooManyMethods")
+public class ReportImageService {
+  private static final String[] ALLOWED_FILETYPES = {"jpg", "png", "gif"};
+
+  @Autowired
+  private ReportImageRepository reportImageRepository;
+
+  /**
```
Comment (palbecki) [inline src/main/java/org/openlmis/report/service/ReportImageService.java:38]: Please correct javdoc

File: src/main/java/org/openlmis/report/web/ReportImageController.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+   * @return ReportImages.
+   */
+  @RequestMapping(method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Iterable<ReportImage> getAllReportImage() {
+    permissionService.canViewReports();
+
+    Iterable<ReportImage> reportImages = reportImageRepository.findAll();
+    if (reportImages == null) {
```
Comment (palbecki) [inline src/main/java/org/openlmis/report/web/ReportImageController.java:89]: Why not return empy iterable?

File: src/main/resources/db/migration/20170801113526552__create_report_Images.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    data bytea,
+    name text NOT NULL
+);
+
+ALTER TABLE ONLY report_images
+    ADD CONSTRAINT report_images_pkey PRIMARY KEY (id);
+
+
+ALTER TABLE ONLY report_images
+    ADD CONSTRAINT uk_1pxqes3xm931yq7evdyfqb5k1 UNIQUE (name);
```
Comment (palbecki) [inline src/main/resources/db/migration/20170801113526552__create_report_Images.sql]: Why don't you use readable constraints' names?

File: src/main/resources/schemas/reportImage.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  "properties": {
+    "id": {
+      "type": "string",
+      "title": "id"
+    },
+    "name": {
+      "type": "string",
+      "title": "name"
+    },
+    "data": {
```
Comment (palbecki) [inline src/main/resources/schemas/reportImage.json:15]: bytes is array of integers?

**Review by palbecki — CHANGES_REQUESTED** [review]

File: src/test/java/org/openlmis/report/service/JasperTemplateServiceTest.java [hunk 8]
```diff
@@ -235,6 +249,38 @@ public void shouldThrowErrorIfTemplateNameAlreadyExists() throws Exception {
   }
 
   @Test
+  public void shouldThrowErrorIfImageDoesNotExist() throws Exception {
+    expectedException.expect(ReportingException.class);
```
Comment (palbecki) [inline src/test/java/org/openlmis/report/service/JasperTemplateServiceTest.java:253]: why is there so much mocking for simple test?

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java (hunk 1 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/domain/JasperTemplate.java]: The table is called jasper_templates

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java (hunk 2 above)
Comment (jkondrat) [inline src/main/java/org/openlmis/report/domain/JasperTemplate.java]: That's what JPA generated, I didn't want to create a third entity explicitly

File: src/main/resources/db/migration/20170801113526552__create_report_Images.sql (hunk 6 above)
Comment (jkondrat) [reply src/main/resources/db/migration/20170801113526552__create_report_Images.sql]: I used auto DDL generation, should I change that anyways?

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java (hunk 1 above)
Comment (palbecki) [reply src/main/java/org/openlmis/report/domain/JasperTemplate.java]: Can't it be changed?

File: src/main/resources/db/migration/20170801113526552__create_report_Images.sql (hunk 6 above)
Comment (palbecki) [reply src/main/resources/db/migration/20170801113526552__create_report_Images.sql]: if you can

File: src/test/java/org/openlmis/report/service/JasperTemplateServiceTest.java (hunk 8 above)
Comment (jkondrat) [reply src/test/java/org/openlmis/report/service/JasperTemplateServiceTest.java:253]: In order to get the parameter and determine whether it's a image, we first have to upload and process the jasper template

File: src/main/java/org/openlmis/report/domain/ReportImage.java (hunk 3 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/domain/ReportImage.java]: Done

File: src/main/java/org/openlmis/report/service/ReportImageService.java (hunk 4 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/service/ReportImageService.java:38]: Done

File: src/main/java/org/openlmis/report/web/ReportImageController.java (hunk 5 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/web/ReportImageController.java:89]: Done

File: src/main/resources/schemas/reportImage.json (hunk 7 above)
Comment (jkondrat) [reply src/main/resources/schemas/reportImage.json:15]: Done

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java (hunk 1 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/domain/JasperTemplate.java]: Done

File: src/main/java/org/openlmis/report/domain/JasperTemplate.java (hunk 2 above)
Comment (jkondrat) [reply src/main/java/org/openlmis/report/domain/JasperTemplate.java]: Done

File: src/main/resources/db/migration/20170801113526552__create_report_Images.sql (hunk 6 above)
Comment (jkondrat) [reply src/main/resources/db/migration/20170801113526552__create_report_Images.sql]: Done

**Pull request merged**